### PR TITLE
fix: repoint dependency packages when Create returns AlreadyExists

### PR DESCRIPTION
**Pull request by erikmiller-gusto**
### Description of your changes

Relocating a package to a new registry (same repository path, new host) silently gets stuck, and a naive fix for that can corrupt an unrelated dependency.

**Relocation gets stuck at `Resolved: True`.** The resolver names a dependency object from its repository path alone (`xpkg.ToDNSLabel(ref.Context().RepositoryStr())`), excluding registry host and digest/tag. When a package moves to a new registry host, the new dependency object collides on that derived name with the existing object, which still points at the old source. `Create` returns `AlreadyExists`, and the resolver has always treated that as success: the existing object is never checked or updated, no error or event is produced, and the Lock's `Resolved` condition stays `True` indefinitely while the dependency is never actually resolved. Since nothing about the Lock changes, nothing re-triggers reconciliation either.

`internal/controller/pkg/revision/dependency.go` already handles the analogous case on the self-registration side — a `PackageRevision` that finds a same-name, different-source lock entry removes it so the package can re-register. The resolver's create-if-missing path had no equivalent.

**The derived name isn't unique.** `ToDNSLabel` drops characters like `_` outright instead of substituting them, and truncates at 63 characters, so two unrelated repository paths can land on the same derived name with no registry relocation involved — `hasheddan/config_nop_c` and `hasheddan/confignopc` both derive to `hasheddan-confignopc`. Repointing on every `AlreadyExists` without checking what it's repointing would silently overwrite an unrelated, already-resolved dependency with a different package.

**This PR, on `AlreadyExists`:**

- **Same repository, different source** — updates the existing object to the desired reference. The comparison is registry-agnostic, so a relocation is recognized as the same dependency.
- **Different repository, same derived name** — leaves the existing object untouched and fails resolution explicitly instead of guessing.
- **Normalizes one comparison edge case** — `RepositoryStr()` implicitly namespaces a bare single-segment repository as `library/x`; without matching that on both sides, a legitimate relocation could misclassify as a collision.

#### Test coverage

`TestReconcileOnNameCollision` (`internal/controller/pkg/resolver/reconciler_test.go`): repoints a stale source, no-ops on an exact match, fails closed on a genuine collision between unrelated packages, and repoints correctly on both sides of the `library/`-namespacing edge case.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Read and followed Crossplane's [AI policy] and confirmed a human stands behind this PR.
- [ ] ~Run `./nix.sh flake check` to ensure this PR is ready for review.~ Nix wasn't available in my environment; ran the equivalent directly instead — `go build`, `go vet`, `go test ./internal/controller/pkg/resolver/...`, and `golangci-lint run ./...` pinned to the flake's resolved version (`v2.12.2`). All clean.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~ Correctness fix to existing dependency-resolution logic, covered by the resolver's existing unit-test harness.
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~ No user-facing or documented behavior change — this restores the resolver's existing dependency-resolution guarantee.
- [ ] Added `backport release-x.y` labels to auto-backport this PR. Leaving to maintainer discretion on which active release branches this should target.
- [ ] ~Followed the [API promotion workflow] if this PR introduces, removes, or promotes an API.~ No API changes.

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[AI policy]: https://github.com/crossplane/crossplane/blob/main/AI_POLICY.md
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[API promotion workflow]: https://github.com/crossplane/crossplane/blob/main/contributing/guide-api-promotion.md

### CODE REVIEW

File: internal/controller/pkg/resolver/reconciler_test.go [hunk 1]
```diff
@@ -1054,6 +1054,197 @@ func TestReconcile(t *testing.T) {
...
+		singleSegmentDependencyPackage = "xpkg.crossplane.io/confignopc"
+		singleSegmentExistingSource    = "confignopc@" + digest2
+	)
+
+	cases := map[string]struct {
+		reason            string
+		dependencyPackage string
+		existingSource    string
+		wantUpdated       bool
+		wantErr           bool
```
[the remark this answers is not in the document]
Comment (erikmiller-gusto) [reply internal/controller/pkg/resolver/reconciler_test.go]: Fixed in 40a7c1a98. `FailsOnUnrelatedCollision` now builds the exact expected error the collision branch constructs and compares it with `cmp.Diff` + `test.EquateErrors()` (this file's existing convention throughout, e.g. lines 1046/1334/1530 - note that's `crossplane-runtime/v2/pkg/test.EquateErrors()`, which compares error type + message, not `cmpopts.EquateErrors()` which does an `errors.Is` check; the latter wouldn't verify anything useful here since this error isn't wrapping a sentinel). Confirmed the weakness was real and the fix catches it: swapped the collision branch's error for an unrelated one and verified the old assertion still passed while the new one fails.

File: internal/controller/pkg/resolver/reconciler.go [hunk 2]
```diff
@@ -398,13 +401,72 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+			// collision is masking a stale, unresolved dependency rather
+			// than reflecting what's actually installed. Repoint the
+			// existing object at the desired reference instead of treating
+			// the AlreadyExists as satisfying the dependency.
+			existing := pack.DeepCopy()
+			if err := r.kube.Get(ctx, types.NamespacedName{Name: pack.GetName()}, existing); err != nil {
+				log.Debug(errGetDependency, "error", err)
+				status.MarkConditions(v1beta1.ResolutionFailed(errors.Wrap(err, errGetDependency)))
+
+				_ = r.kube.Status().Update(ctx, lock)
```
[the remark this answers is not in the document]
Comment (erikmiller-gusto) [reply internal/controller/pkg/resolver/reconciler.go:430]: Declining this one, evaluated on the merits: the branches this PR adds (Get/Update failures at what's now ~426-433 and ~461-468) mark `ResolutionFailed` on any client error, transient or not - true. But that's not a new inconsistency this PR introduces; it's the pre-existing convention for every other client-error branch in this same function, including code this PR doesn't touch: the `r.kube.List(ctx, l)` failure a few lines above (originally ~334-341) and the adjacent, pre-existing `Create` error branch immediately above the code I added (originally ~404-411, present before this PR) both mark `ResolutionFailed` unconditionally on any error from those calls too.\n\nSingling out only the two new branches would leave the function internally inconsistent rather than resolve the inconsistency, and a proper fix (distinguishing retryable client errors from deterministic ones) would mean auditing every error path in `Reconcile`, which is a separate, pre-existing architectural question and out of scope for a fix targeted at the AlreadyExists collision bug. It's also worth noting the condition isn't persistently stuck the way a stale spec would be: `MarkConditions` overwrites it every reconcile, and the non-nil return triggers a requeue, so a transient Get/Update failure here self-corrects on the very next attempt exactly like the pre-existing List/Create branches already do.\n\nHappy to take this up as a follow-up PR auditing transient-vs-deterministic error handling across the whole function if that's valuable, but treating it as a merge blocker for this PR specifically doesn't hold up once the surrounding code is considered.

**Comment by erikmiller-gusto** [comment]
Re: the walkthrough's risk summary ("an upgrade path can still repoint an existing package using only its normalized image name, potentially affecting an unrelated package and its dependents") - I looked into this since it wasn't one of the two actionable inline comments, but wanted to make sure it wasn't a real gap. It isn't: the upgrade path (a few lines above what this PR changes, gated behind `EnableAlphaDependencyVersionUpgrades`) finds the existing package via `r.kube.List` over every installed package of that kind and filters with `pref.Context().Name() == ref.Context().Name()` - a full, exact registry+repository string match using `name.StrictValidation`, not the lossy `xpkg.ToDNSLabel`-derived name this PR's collision guard is about. It can't collide the way the `Create`/`AlreadyExists` path can, since it never relies on the truncated/underscore-dropping derived name at all. No change needed there.

Addressed the two actionable comments individually inline: fixed the test one in 40a7c1a98, and replied on the transient-error one explaining why I don't think it's a regression specific to this PR (same pattern predates it in the immediately adjacent `Create` and `List` error branches).

**Review by Karthik-Chowdary — APPROVED** [review]
Reviewed the full resolver path at 40a7c1a9, including the create/AlreadyExists branch, registry-agnostic repository comparison, implicit library/ normalization, collision fail-closed behavior, and the exact-error regression update. The implementation only updates an existing dependency when it is the same normalized repository, leaves unrelated derived-name collisions untouched, and preserves the no-op path for an exact source match. I also ran `go test ./internal/controller/pkg/resolver` and the focused `TestReconcileOnNameCollision` test; both pass. No blocking issues found.
